File: src/medium/ThreeSum.java
package medium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// https://leetcode.com/problems/3sum/
public class ThreeSum {
  public List<List<Integer>> threeSum(int[] nums) {
    List<List<Integer>> result = new ArrayList<>();
    if (nums == null) {
      return result;
    }
    Arrays.sort(nums);  // O(nlogn)
    for (int i = 0; i < nums.length - 2; i++) { // O(n2)
      // skip dups
      if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
        int l = i + 1;
        int r = nums.length - 1;
        int t = 0 - nums[i];
        while (l < r) {
          if (nums[l] + nums[r] == t) {
            List<Integer> list = new ArrayList<>();
            list.add(nums[i]);
            list.add(nums[l]);
            list.add(nums[r]);
            result.add(list);
            while (l < r && nums[l] == nums[l + 1]) l++;    // skip dups
            while (l < r && nums[r] == nums[r - 1]) r--;    // skip dups
            l++;
            r--;
          } else if (nums[l] + nums[r] > t) {
            r--;
          } else {
            l++;
          }
        }
      }
    }
    return result;
  }
}

File: csharp/LeetCode/Medium/ThreeSum.cs
namespace LeetCode.Medium.ThreeSum
{
    using System;
    using System.Collections.Generic;

    public class Solution
    {
        public IList<IList<int>> ThreeSum(int[] nums)
        {
            Array.Sort(nums);
            var ans = new List<IList<int>>();
            for (var i = 0; i < nums.Length; i++)
            {
                if (i > 0 && nums[i] == nums[i - 1])
                    continue;
                var target = -nums[i];
                var l = i + 1;
                var r = nums.Length - 1;
                while (l < r)
                {
                    var twoSum = nums[l] + nums[r];
                    if (twoSum > target)
                    {
                        r--;
                    }
                    else if (twoSum < target)
                    {
                        l++;
                    }
                    else
                    {
                        ans.Add(new int[] { -target, nums[l], nums[r] });
                        do
                        {
                            l++;
                        }
                        while (l < r && nums[l] == nums[l - 1]);
                        do
                        {
                            r--;
                        }
                        while (l < r && nums[r] == nums[r + 1]);
                    }
                }


            }
            return ans;
        }
    }
}
